### ISSUE TITLE
The wrong http settings may be used if there are two git repo configs for the same url

### ISSUE THREAD

**Issue by marnee01** [spring-cloud/spring-cloud-config#1482]
While investigating issue [Error: "More than one git repo URL template matched URL"](https://github.com/spring-cloud/spring-cloud-config/issues/1480), I spotted another possible defect in the HttpClientConfigurableHttpConnectionFactory implementation.

If someone has multiple git repo configs for the same url, but with different http settings, only one of the http settings will be used, regardless of how the request coming in. 

Note: This does not impact my team, but since I spotted it, I thought I would relay this.

If someone has a configuration such as:

```
spring:
  cloud:
    config:
      server:
        git:
          uri: https://blahblah/v1/repos/vvv_configs-{application}
          proxy:
            https:
              host: ${proxyuri}
              port: '8080'

          repos:

           # No proxy in dev environment
            WebApp-Dev:
              pattern: webapp/dev
              searchPaths: SharedDomains/*, {application}
              basedir: ${basedirParent}\\baseWebApp
              uri: https://blahblah/v1/repos/vvv_configs-webapp
				  
           # Proxy needed for all environments except dev
            WebApp:
              pattern: webapp
              searchPaths: SharedDomains/*, {application}
              basedir: ${basedirParent}\\baseWebApp
              uri: https://blahblah/v1/repos/vvv_configs-webapp
              proxy:
                https:
                  host: ${proxyuri}
                  port: '8080'
```

The `HttpClientConfigurableHttpConnectionFactory.addConfiguration` method puts http clients into its internal map with the url as key. Since url is the same, that means that there will only be one entry for the vvv_configs-webapp URL in the map, even though there are two different configs for that URL. Whether the http client will be the one using the proxy or the one without is indeterminate (second one added to the map wins, obviously).

I did a quick test, and confirmed this behavior. Whether I call with "dev" profile or not, it will always use the same http client object. I am using Greenwich SR3.

**Comment by marnee01** [comment]
I spoke too soon about this not impacting us. We have a new requirement where we don't want to use a proxy when client app is running in Cloud. So, we'd like it to pass a "noproxy" profile and have a separate repo definition that doesn't contain proxy info. (I don't see any other way to disable proxy dynamically.) 

So, if you just replaced "pattern: webapp/dev" with "pattern: webapp/noproxy" in the config I referenced in the comment above, that is what we basically want to do.

I will attach a project that demonstrates this tomorrow.

**Comment by marnee01** [comment]
I created a sample project for the issue: https://github.com/marnee01/vts-spring-config-server-1482.

Instructions are in the README file at the root of the project.

**Comment by marnee01** [comment]
A workaround or way to avoid this problem that is working for us right now is to switch to using the https.proxyHost/Port system properties, rather than adding proxy settings for each git repo. (And for the Cloud ConfigServer instance, we just won't add those properties.) That turns out to be a better approach for us, anyway.

That might not be feasible for some scenarios, but wanted to mention it for anyone else reading this who is looking for a workaround.

**Comment by ryanjbaxter** [comment]
When I apply the fix from #1480 that does not seem to address this issue, there is still only 2 entries in the map.  I don't see how that would address the uniqueness of the key.

**Comment by marnee01** [comment]
If there are two connections that would match an incoming URL, the original code would find that there are two matches, log an error, and not use either one.

What the workaround does is that it chooses the first matching connection that doesn't have a placeholder, if any. If there is one and only one such connection, that is the one used. It ignores any connection containing a placeholder that would match. By contrast, the original code would match against all connections, including those with placeholders, and if there was more than one match, it would log an error and use neither.

The relevant code is the following. The first line looks up matches against connections with no placeholders (since second argument is false):

```
		Map<String, HttpClientBuilder> builderMap = lookupHttpClientBuilders(url, false);

		// Only if no match, try for URLs with placeholders.
		if (builderMap.isEmpty()) {
			// This is the original logic:
			builderMap = this.lookupHttpClientBuilders(url, true);
		}
```
I really don't see this as a "fix", just a workaround.

**Comment by ryanjbaxter** [comment]
OK got it so since one is selected we instead of 2 being returned that would solve the problem.  Thanks!

**Comment by ryanjbaxter** [comment]
Closed via  c7ab0f3

**Comment by marnee01** [comment]
Please re-visit and possibly re-open this. This was marked as fixed, but I don't see how it could be.

Ryan Baxter had said:

When I apply the fix from #1480 that does not seem to address this issue, there is still only 2 entries in the map. I don't see how that would address the uniqueness of the key.

I replied to that, but I actually made a mistake and thought I was replying on the issue documented in #1480. (So, my response starting with "If there are two connections that would..." from Nov 15th is not applicable.) Ryan was in fact correct in saying that the fix from 1480 won't fix this. I am very sorry for that confusion.

In fact, it appears to me that there is an underlying design flaw in this area (which is why I kept insisting the workaround for 1480 was nothing but a workaround). The code in question does not take into account that you can have the same URL but with different properties provided in the spring.cloud.config.server.git properties. The code knows nothing about the spring.cloud.config.server.git configuration.
